### PR TITLE
util/decoder: migrate test-infra to testify

### LINKED ISSUE #26414: migrate test-infra to testify for `util/decoder` pkg

### PR DESCRIPTION
**Pull request by feitian124**
### What problem does this PR solve?

Issue Number: close #26414 

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by feitian124** [comment]
/cc @tisonkun
please help review, thanks

File: util/rowDecoder/decoder_test.go [hunk 1]
```diff
@@ -202,18 +194,27 @@ func (s *testDecoderSuite) TestClusterIndexRowDecoder(c *C) {
...
-			c.Assert(err1, IsNil)
-			c.Assert(equal, Equals, 0)
+			require.Nil(t, err1)
+			require.Equal(t, 0, equal)
 		}
 	}
 }
+
+// TODO: local copy of testutil.MustNewCommonHandle，should use function in testutil after it migrate to testify
+func MustNewCommonHandle(t *testing.T, values ...interface{}) kv.Handle {
```
Comment (tisonkun) [inline util/rowDecoder/decoder_test.go:214]: You can just create a file named, for example, "handle.go" under "testkit" and do the todo a favor.

**Review by tisonkun — CHANGES_REQUESTED** [review]
Generally looks good. Comment inline that you can extract the method in this pass. Ask me for help if you need it.

File: util/rowDecoder/decoder_test.go (hunk 1 above)
Comment (feitian124) [reply util/rowDecoder/decoder_test.go:214]: sure, a little busy now hope finish before bed

Comment (tisonkun) [reply util/rowDecoder/decoder_test.go:214]: Thanks ❤️

Comment (feitian124) [reply util/rowDecoder/decoder_test.go]: sorry, i met some problems and not sure how to resolve.
1. decoder_test.go unit test failed for goleak
```shell
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 21 in state select, with go.opencensus.io/stats/view.(*worker).start on top of the stack:
goroutine 21 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc000162410)
	/home/ming/go/pkg/mod/go.opencensus.io@v0.22.3/stats/view/worker.go:154 +0xcd
created by go.opencensus.io/stats/view.init.0
	/home/ming/go/pkg/mod/go.opencensus.io@v0.22.3/stats/view/worker.go:32 +0x57
```

2. key_test.go and keydecoder_test.go have cycle import with util/testkit/handle.go
```shell
# github.com/pingcap/tidb/kv
FAIL	github.com/pingcap/tidb/kv [setup failed]
package github.com/pingcap/tidb/kv (test)
	imports github.com/pingcap/tidb/util/testkit
	imports github.com/pingcap/tidb/domain
FAIL
	imports github.com/pingcap/tidb/bindinfo
	imports github.com/pingcap/tidb/sessionctx
	imports github.com/pingcap/tidb/kv: import cycle not allowed in test

```

Comment (feitian124) [reply util/rowDecoder/decoder_test.go]: i need some dig to resolve the 2 problems, any suggestion is appreciated

Comment (tisonkun) [reply util/rowDecoder/decoder_test.go]: For goleak, it is a known goroutine that doesn't exit, you can ignore by

```go
        opts := []goleak.Option{
		goleak.IgnoreTopFunction("go.etcd.io/etcd/pkg/logutil.(*MergeLogger).outputLoop"),
		goleak.IgnoreTopFunction("go.opencensus.io/stats/view.(*worker).start"),
	}
	goleak.VerifyTestMain(m, opts...)
```

for cycle import, you can try to change `package keydecoder` to `package keydecoder_test`

Comment (feitian124) [reply util/rowDecoder/decoder_test.go]: thanks, helps a lot

Comment (feitian124) [reply util/rowDecoder/decoder_test.go]: updated, please review again, thanks

File: util/testkit/handle.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	"testing"
+
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/stretchr/testify/require"
+)
+
+func MustNewCommonHandle(t *testing.T, values ...interface{}) kv.Handle {
```
Comment (tisonkun) [inline util/testkit/handle.go:13]: You should add under "testkit" instead of "util/testkit". I will push a follow up commit.

Comment (tisonkun) [reply util/testkit/handle.go]: Done https://github.com/pingcap/tidb/pull/26629/commits/7c328357797fe42536c8bbc5ca526279a52c944f. Also be aware of License header which is required by APL 2.0.

**Review by tisonkun — APPROVED** [review]
LGTM

File: util/testkit/handle.go (hunk 2 above)
Comment (feitian124) [reply util/testkit/handle.go]: ok, thanks.

**Comment by feitian124** [comment]
hello @tisonkun , anything else need me to do to merge this pr?

**Comment by tisonkun** [comment]
@feitian124 thanks for ping me. I'll ping other reviewers for you.

/cc @xhebox @tiancaiamao

**Review by xhebox — APPROVED** [review]

**Review by mmyj — APPROVED** [review]
lgtm

**Comment by tisonkun** [comment]
@mmyj may you help on merging as there are 2 LGTM2?

**Comment by feitian124** [comment]
ci failed, i checked a few latest pr merged to this branch and try run unit test:
- statistics/handle/update_test.go 
```shell
OOPS: 109 passed, 1 skipped, 3 FAILED, 1 PANICKED
--- FAIL: TestT (24.31s)
```

- statistics/selectivity_test.go
```shell
/home/ming/learn/go/pingcap_projects/tidb/util/testkit/testkit.go:63:
    res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)
... obtained string = "" +
...     "[IndexReader_6 1.36 root  index:IndexRangeScan_5]\n" +
...     "[└─IndexRangeScan_5 1.36 cop[tikv] table:exp_backoff, index:idx(a, b, c, d) range:[1 1 1 3,1 1 1 5], keep order:false]\n"
... expected string = "" +
...     "[IndexReader_6 0.00 root  index:IndexRangeScan_5]\n" +
...     "[└─IndexRangeScan_5 0.00 cop[tikv] table:exp_backoff, index:idx(a, b, c, d) range:[1 1 1 3,1 1 1 5], keep order:false]\n"
... sql:explain select * from exp_backoff where a = 1 and b = 1 and c = 1 and d >= 3 and d<= 5, args:[]

OOPS: 58 passed, 1 FAILED
--- FAIL: TestT (7.62s)
```

**Comment by feitian124** [comment]
my change should have no impact on other package, not sure why they are fail.
any idea?
@tisonkun

**Comment by tisonkun** [comment]
@feitian124 you can check #25899 too see if there is a ticket for this test.

**Comment by feitian124** [comment]
> @feitian124 you can check #25899 to see if there is a ticket for this test.

not found, may be need add the above 2.
it's surprising to find there are so many unstable testing, no wonder i run the unit twice, the failed test count is not equal..

**Comment by tisonkun** [comment]
> > @feitian124 you can check #25899 to see if there is a ticket for this test.
> 
> not found, may be need add the above 2.
> it's surprising to find there are so many unstable testing, no wonder i run the unit twice, the failed test count is not equal..

Yes. It is one thing our community struggle for...hopefully we converge with the subtasks of tracking issue done one by one and you can digest some of them when migrate tests.

**Pull request merged**